lib/source_file.cpp

```diff
@@ -168,7 +168,12 @@ SourceFileLine InputSources::getSourceLine(unsigned line) const {
     // ----------
     // The first line indicates that line 2 is the first line in x.p4
     // line=2, it->first=1, it->second.sourceLine=1
-    // So we have to subtract one to get the real line number
+    // So we have to subtract one to get the real line number,
+    // except in the special case of all 0s, where 0 is preferable to
+    // (-1 mod 2^32)
+    if (line == 0 && it->first == 0 && it->second.sourceLine == 0) {
+        return SourceFileLine(it->second.fileName, 0);
```

Comment:
If you compile this program: https://github.com/jafingerhut/p4-guide/blob/master/demo5/demo5.p4_16.p4#L89

using the command `p4c-bm2-ss -o tmp.json demo5.p4_16.p4`, the action `top_level_action`, being defined at the top level, and having an `inout` struct of all headers, but only modifying one field of one header inside of that struct, is compiled to contain several `copy_header` calls synthesized by the compiler to implement the copy-in and copy-out semantics.  One of those has source info of 2^32-1 in the output, which becomes 0 if you make the diff suggested in this PR.

This might very well be a bug elsewhere in the compiler, but I haven't tried to trace it backwards from the point of this change.